tests: Change the result compare method in pin_get_config()

hunterhu3000 opened the pull request:
Fixes #47921
When setting a GPIO to certain value, and then get the GPIO value, on some
platforms, such as ITE, these two values may not be the same. For example
setting GPIO_OUTPUT | GPIO_OUTPUT_INT_HIGH, it will read back
GPIO_OUTPUT | GPIO_OUTPUT_INT_HIGH | GPIO_VOLTAGE_3P3. This is as expected,
as GPIO_VOLTAGE is a read only property. So when comparing the gpio set
and get values, it is necessary to filter the read only bits first and
then compare to the set value.

Signed-off-by: Hu Zhenyu <zhenyu.hu@intel.com>

hunterhu3000 commented:
This PR is to fix https://github.com/zephyrproject-rtos/zephyr/issues/47921

henrikbrixandersen reviewed (changes requested):
Please change the Git commit log to describe the actual changes.

FRASTM commented:
This fixes the pb as reported by issues https://github.com/zephyrproject-rtos/zephyr/issues/47921 and https://github.com/zephyrproject-rtos/zephyr/issues/48007

In any case the test should also PASS when  `CONFIG_GPIO_GET_CONFIG=n`
--> flag `pin_get_config() `with `#ifdef CONFIG_GPIO_GET_CONFIG`

MaureenHelm commented:
> This PR is to fix #47921

Please add "Fixes #47921" to the PR description. This will link the Issue and close it automatically when the PR gets merged.
https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

FRASTM commented:
When the function to get the gpio pin config was introduced by the  PR #43225, flags was supposed to give the position of the gpio : in/out GPIO_OPEN_DRAIN/ GPIO_PULL_DOWN/GPIO_PULL_UP
I guess this information should fit the [GPIO input/output configuration flags](https://docs.zephyrproject.org/latest/hardware/peripherals/gpio.html?highlight=gpio_flags_t#overview) 
If the purpose is to know about high or low states only of the gpio, what is the benefit of gpio_stm32_get_config() function when reading the 
```
	otype = LL_GPIO_GetPinOutputType(gpio, pin_ll);
	pupd = LL_GPIO_GetPinPull(gpio, pin_ll);
	mode = LL_GPIO_GetPinMode(gpio, pin_ll);
```

mmahadevan108 commented:
cc @semihalf-barnas-michal

semihalf-barnas-michal commented:
> If the purpose is to know about high or low states only of the gpio, what is the benefit of gpio_stm32_get_config() function

Exactly, the gpio_get_config should return all (supported) flags set for the gpio, not only input or output. So this PR hides the problem with the stm32 driver instead of fixing it.
There is probably some problem in the implementation of stm32 driver, some flags are missing or are reading invalid enums for specific stm boards.

> In any case the test should also PASS when `CONFIG_GPIO_GET_CONFIG=n`
> suggestion : flag `pin_get_config() `with `#ifdef CONFIG_GPIO_GET_CONFIG`

Or just wrap the whole test with the ifdef. It may be better to not have this test run, when it's not implemented, instead of returning the positive.

hunterhu3000 commented:
> This fixes the pb as reported by issues #47921 and #48007
> 
> In any case the test should also PASS when `CONFIG_GPIO_GET_CONFIG=n` suggestion : flag `pin_get_config() `with `#ifdef CONFIG_GPIO_GET_CONFIG`
OK

hunterhu3000 commented:
> gpio_stm32_get_config

I think the user can use pin_get_config to get more infomation than just the high/low input/output. But some of the infomation is read only and may not be used in gpio_pin_configure. So in the test, it is enough to check the high/low input/output.

hunterhu3000 commented:
> > If the purpose is to know about high or low states only of the gpio, what is the benefit of gpio_stm32_get_config() function
> 
> Exactly, the gpio_get_config should return all (supported) flags set for the gpio, not only input or output. So this PR hides the problem with the stm32 driver instead of fixing it. There is probably some problem in the implementation of stm32 driver, some flags are missing or are reading invalid enums for specific stm boards.
> 
> > In any case the test should also PASS when `CONFIG_GPIO_GET_CONFIG=n`
> > suggestion : flag `pin_get_config() `with `#ifdef CONFIG_GPIO_GET_CONFIG`
> 
> Or just wrap the whole test with the ifdef. It may be better to not have this test run, when it's not implemented, instead of returning the positive.

I think #48007 is another issue. This PR will not fix it, as the set value of #48007 is 0xa0000 but the get value is 0x20000b74. 0xa0000 & 0x20000b74 is 0, so the assertion in this PR cannot pass. For #48007, we need to check gpio_stm32_get_config(), it seems does not follow the zephyr definition

semihalf-barnas-michal commented:
> I think #48007 is another issue

Yes, there seems to be a bug in the driver implementation. Both issues are connected to my PR, but definitely different root cause.

semihalf-barnas-michal commented:
> I will modify my patch to avoid this case

Thanks

semihalf-barnas-michal commented:
That looks great

cfriedt reviewed (dismissed):
LGTM

FRASTM commented on tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c:
```diff
@@ -633,11 +633,17 @@ static int bits_logical(void)
 
 /* gpio_pin_get_config()
  */
+#ifdef CONFIG_GPIO_GET_CONFIG
```
Better include the comment line in the 
#ifdef CONFIG_GPIO_GET_CONFIG

FRASTM reviewed (commented):
else LGTM

mmahadevan108 commented:
> > This PR is to fix #47921
> 
> Please add "Fixes #47921" to the PR description. This will link the Issue and close it automatically when the PR gets merged. https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

@hunterhu3000 as suggested earlier please add `Fixes #47921` to the description. This will then automatically close the issue when this PR gets merged.

henrikbrixandersen commented on tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c:
```diff
@@ -633,11 +633,17 @@ static int bits_logical(void)
 	gpio_flags_t flags_get = 0;
 	gpio_flags_t flags_set;
 	int rc;
+	gpio_flags_t flags_mask = GPIO_INT_MASK | GPIO_OUTPUT_INIT_LOGICAL |
+				  GPIO_OUTPUT_INIT_HIGH | GPIO_OUTPUT_INIT_LOW |
+				  GPIO_OUTPUT | GPIO_INPUT |
+				  GPIO_PULL_UP | GPIO_PULL_DOWN |
+				  GPIO_SINGLE_ENDED | GPIO_LINE_OPEN_DRAIN | GPIO_ACTIVE_LOW;
```
So the idea is to mask out SoC-specific GPIO flags? Why not just use `GENMASK` to generate a bit mask excluding those? The above would need to be updated every time we add a new flag?

```
/**
 * @brief Provides a type to hold GPIO devicetree flags.
 *
 * All GPIO flags that can be expressed in devicetree fit in the low 16
 * bits of the full flags field, so use a reduced-size type to record
 * that part of a GPIOS property.
 *
 * The lower 8 bits are used for standard flags. The upper 8 bits are reserved
 * for SoC specific flags.
 */
typedef uint16_t gpio_dt_flags_t;

/**
 * @brief Provides a type to hold GPIO configuration flags.
 *
 * This type is sufficient to hold all flags used to control GPIO
 * configuration, whether pin or interrupt.
 */
typedef uint32_t gpio_flags_t;

```

semihalf-barnas-michal replied:
> So the idea is to mask out SoC-specific GPIO flags?

In this case the problem was with voltage flags from device-tree. If you don't specify any voltage (so flag is GPIO_VOLTAGE_DEFAULT), you can read a real voltage from board registers (GPIO_VOLTAGE_3P3).

hunterhu3000 replied:
OK, use GENMASK

hunterhu3000 commented:
@henrikbrixandersen Could you please review my latest patch? If if is OK, thanks to +1.

hunterhu3000 commented on tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c:
```diff
@@ -672,6 +683,8 @@ void test_gpio_port(void)
 		      "bits_logical failed");
 	zassert_equal(check_pulls(), TC_PASS,
 		      "check_pulls failed");
+#ifdef CONFIG_GPIO_GET_CONFIG
```
Updated. Thanks to review

FRASTM commented on tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c:
```diff
@@ -672,6 +683,7 @@ void test_gpio_port(void)
-	zassert_equal(pin_get_config(), TC_PASS,
-		      "pin_get_config failed");
+	if (IS_ENABLED(CONFIG_GPIO_GET_CONFIG))
+		zassert_equal(pin_get_config(), TC_PASS,
```
In case of CONFIG_GPIO_GET_CONFIG=n, the pin_get_config might definitively be not supported in the driver, then giving compilation warning --> function is undeclared 
This is not exactly the same as 
```
#ifdef CONFIG_GPIO_GET_CONFIG
 	zassert_equal(pin_get_config(), TC_PASS,
 		      "pin_get_config failed");
#endif /* CONFIG_GPIO_GET_CONFIG */
```

hunterhu3000 replied:
Yes, we'd better to roll back

cfriedt replied:
Should be `IS_ENABLED(..) && api->get_config != NULL`

hunterhu3000 replied:
> pin_get_config
Updated. It seems the CONFIG_GPIO_GET_CONFIG for the function pin_get_config should be remove as well

FRASTM commented:
BTW if you could please include the change in the [tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c](https://github.com/zephyrproject-rtos/zephyr/pull/48159/commits/717e6d26f7d57264def8fe67a956b864bd3e7735#diff-0cf45b2fd058a429827c4bfe9c9602d63aa2cce9b39e3779a8bfb354ba08e331)
to handle the correct error code from the gpio_pin_get_config return
```
rc = gpio_pin_get_config(dev, PIN_OUT, &flags_get);
	if (rc == -ENOSYS) {
		return TC_PASS;
	}
```

hunterhu3000 commented:
> BTW if you could please include the change in the [tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c](https://github.com/zephyrproject-rtos/zephyr/pull/48159/commits/717e6d26f7d57264def8fe67a956b864bd3e7735#diff-0cf45b2fd058a429827c4bfe9c9602d63aa2cce9b39e3779a8bfb354ba08e331) to handle the correct error code from the gpio_pin_get_config return
> 
> ```
> rc = gpio_pin_get_config(dev, PIN_OUT, &flags_get);
> 	if (rc == -ENOSYS) {
> 		return TC_PASS;
> 	}
> ```

I have another question, why ENOSYS is regarded as PASS？

FRASTM commented:
> > BTW if you could please include the change in the [tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c](https://github.com/zephyrproject-rtos/zephyr/pull/48159/commits/717e6d26f7d57264def8fe67a956b864bd3e7735#diff-0cf45b2fd058a429827c4bfe9c9602d63aa2cce9b39e3779a8bfb354ba08e331) to handle the correct error code from the gpio_pin_get_config return
> > ```
> > rc = gpio_pin_get_config(dev, PIN_OUT, &flags_get);
> > 	if (rc == -ENOSYS) {
> > 		return TC_PASS;
> > 	}
> > ```
> 
> I have another question, why ENOSYS is regarded as PASS？

The reason for this is in the [gpio_pin_get_config](https://docs.zephyrproject.org/latest/doxygen/html/group__gpio__interface.html#ga133284b60a13c6dc50b87c1d04259f5e)
 The gpio_pin_get_config()  [zephyr/include/zephyr/drivers/gpio.h](https://github.com/zephyrproject-rtos/zephyr/blob/352fde1271c568fa24cb48b4aed20f97b25945b1/include/zephyr/drivers/gpio.h#L881-L883) return code is _not_ ENOTSUP 
```
	if (api->pin_get_config == NULL)
		return -ENOSYS;
```

hunterhu3000 commented:
If getting current pin configuration is not implemented by the driver, then we should not run this test. Returning fail will remind us there might be a  problem or not running the test next time if it is as expected, while returning pass will cover the problem. @FRASTM

cfriedt commented on tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c:
```diff
@@ -633,27 +633,35 @@ static int bits_logical(void)
 	zassert_equal(rc, 0, "pin configure failed");
 
 	rc = gpio_pin_get_config(dev, PIN_OUT, &flags_get);
-	if (rc == -ENOTSUP) {
```
This should probably be kept. The reason is that there could be multiple GPIO drivers and one of them might support this feature, while the other does not.

That's the main reason why I kept the `gpio_get_direction()` tests in a separate suite.

hunterhu3000 replied:
Only one GPIO driver gpio_stm32_get_config() will return ENOTSUP. When the runtime pm is not defined, the function will return ENOTSUP, and the gpio_pin_configure will not be performed. To my understanding, at this time, the test case should be regarded as TC_FAIL. Do you agree?

hunterhu3000 replied:
This part is modified in https://github.com/zephyrproject-rtos/zephyr/pull/48159/files. So in this PR, I will not touch this.

cfriedt reviewed (changes requested):
Sorry to ask for another rev. Please see my comment above.

hunterhu3000 commented:
Rebased to the latest. Thanks to review

hunterhu3000 commented:
As there is a long history of this PR, I'd like to summarize it below:
1. henrikbrixandersen/MaureenHelm/mmahadevan108 asked to change the commit log -> done
2. FRASTM think this PR can fix #48007, but they are two different problems. #48077 is fixed by #48159 later.
3. semihalf-barnas-michal pointed out that the patch cannot cover the corner case such as 0xFFFFFFFF -> fixed
4. semihalf-barnas-michal to add "#ifdef CONFIG_GPIO_GET_CONFIG" -> done
5. henrikbrixandersen asked to use GENMASK instead of OR -> done
6. cfriedt asked to use "if (IS_ENABLED(CONFIG_..))" instead of "#ifdef CONFIG_GPIO_GET_CONFIG" -> done
7. FRASTM asked to chaged the ENOTSUP to ENOSYS -> cfriedt disagree this. So nothing is done on this PR. The modification was done in #48159

hunterhu3000 commented:
@mnkp @FRASTM @MaureenHelm @henrikbrixandersen @cfriedt @semihalf-barnas-michal 
Could you please help to review and +1? Thanks.

hunterhu3000 commented on tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c:
```diff
@@ -639,6 +639,16 @@ static int pin_get_config(void)
+	 * GPIO_SINGLE_ENDED/PUSH_PULL	bit 1
+	 * GPIO_ACTIVE_LOW/HIGH		bit 0
+	 */
+	gpio_flags_t flags_mask = GENMASK(26, 16) | GENMASK(5, 4) | GENMASK(2, 0);
```
These values cannot be used directly, as they are bits, not direct values, eg.
#define GPIO_INT_HIGH_1       (1U << 26)
#define GPIO_PULL_UP            (1 << 4)

henrikbrixandersen replied:
Isn't the purpose to exclude all vendor/SoC specific flags? How about just `GENMASK(31, 16) | GENMASK(7, 0)`?

cfriedt replied:
Alternatively, just explicitly define macros used in the variable by or'ing them all together. I'm sure that was the first rev. I agree with what Brix says as well though, this should only include non-vendor-specific flags.

hunterhu3000 replied:
The bits in the flags is discontinuous，eg bit 26~31 is not defined and maybe used some day in the future. If we use GENMASK(31,16), this test case may fail some day, if bit 31 is defined as a read only property.
```
#define GPIO_ACTIVE_LOW         (1 << 0)
#define GPIO_ACTIVE_HIGH        (0 << 0)
#define GPIO_SINGLE_ENDED       (1 << 1)
#define GPIO_PUSH_PULL          (0 << 1)
#define GPIO_LINE_OPEN_DRAIN    (1 << 2)
#define GPIO_LINE_OPEN_SOURCE   (0 << 2)
#define GPIO_PULL_UP            (1 << 4)
#define GPIO_PULL_DOWN          (1 << 5)

#define GPIO_VOLTAGE_POS        11
#define GPIO_VOLTAGE_MASK       (3U << GPIO_VOLTAGE_POS)
#define GPIO_VOLTAGE_DEFAULT    (0U << GPIO_VOLTAGE_POS)
#define GPIO_VOLTAGE_1P8        (1U << GPIO_VOLTAGE_POS)
#define GPIO_VOLTAGE_3P3        (2U << GPIO_VOLTAGE_POS)
#define GPIO_VOLTAGE_5P0        (3U << GPIO_VOLTAGE_POS)

#define GPIO_INPUT              (1U << 16)
#define GPIO_OUTPUT             (1U << 17)
#define GPIO_OUTPUT_INIT_LOW    (1U << 18)
#define GPIO_OUTPUT_INIT_HIGH   (1U << 19)
#define GPIO_OUTPUT_INIT_LOGICAL (1U << 20)

#define GPIO_INT_DISABLE               (1U << 21)
#define GPIO_INT_ENABLE                (1U << 22)
#define GPIO_INT_LEVELS_LOGICAL        (1U << 23)
#define GPIO_INT_EDGE                  (1U << 24)
#define GPIO_INT_LOW_0                 (1U << 25)
#define GPIO_INT_HIGH_1                (1U << 26)
```
I agree it is not good to use the magic numbers. If so we need to double he macro numbers and define like this:
```
#define GPIO_ACTIVE_SHIFT       0
#define GPIO_PULL_UP_SHIFT     4
GENMASK(GPIO_INT_HIGH_1_SHIFT, GPIO_INPUT_SHIFT)
```
I think it is also not friendly to the code reader. As the reader still does not know what is between GPIO_INT_HIGH_1_SHIFT and GPIO_INPUT_SHIFT.
Maybe using the first version(bit or, not the GENMASK) is better.

cfriedt replied:
I like that GPIO_VOLTAGE_MASK exists. Couldn't we just do something similar for padconf, Input-Output, and int

hunterhu3000 replied:
I have uploaded a new version, is that what you mean?

hunterhu3000 replied:
@cfriedt could you please help to review this patch? Thanks

hunterhu3000 committed:
tests: Change the result compare method in pin_get_config()

Fixes #47921
When setting a GPIO to certain value, and then get the GPIO value, on some
platforms, such as ITE, these two values may not be the same. For example
setting GPIO_OUTPUT | GPIO_OUTPUT_INT_HIGH, it will read back
GPIO_OUTPUT | GPIO_OUTPUT_INT_HIGH | GPIO_VOLTAGE_3P3. This is as expected,
as GPIO_VOLTAGE is a read only property. So when comparing the gpio set
and get values, it is necessary to filter the read only bits first and
then compare to the set value.

Signed-off-by: Hu Zhenyu <zhenyu.hu@intel.com>

hunterhu3000 commented:
@henrikbrixandersen Could you please review and +1 so that the PR can be merged to the mainline earlier, thanks.
